src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs

```diff
@@ -22,16 +22,20 @@ public class DefaultConnectionContext : ConnectionContext,
                                             IConnectionHeartbeatFeature,
                                             ITransferFormatFeature
     {
-        private List<(Action<object> handler, object state)> _heartbeatHandlers = new List<(Action<object> handler, object state)>();
+        private object _heartBeatLock = new object();
+        private List<(Action<object> handler, object state)> _heartbeatHandlers;
 
         // This tcs exists so that multiple calls to DisposeAsync all wait asynchronously
         // on the same task
         private TaskCompletionSource<object> _disposeTcs = new TaskCompletionSource<object>();
 
-        public DefaultConnectionContext(string id, IDuplexPipe transport, IDuplexPipe application)
+        /// <summary>
+        /// Creates the DefaultConnectionContext without Pipes to avoid upfront allocations.
+        /// You'll need to add Transport and Application pipes on your own.
```

Comment:
~~You'll need to add Transport and Application pipes on your own~~

`The caller is expected to set the <see cref="Transport"/> and <see cref="Application"/> pipes manually`